src/MOI_wrapper.jl

```diff
@@ -0,0 +1,225 @@
+end
+
+function MOI.empty!(model::Optimizer)
+    model.num_variables = 0
+    model.manifold = nothing
+    empty!(model.variable_primal_start)
+    model.sense = MOI.FEASIBILITY_SENSE
+    model.nlp_data = MOI.NLPBlockData([], _EmptyNLPEvaluator(), false)
+    model.qp_data = QPBlockData{Float64}()
+    return nothing
+end
+
+MOI.get(::Optimizer, ::MOI.SolverName) = "Manopt"
+
+MOI.supports_incremental_interface(::Optimizer) = true
+
+function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike)
+    return MOI.default_copy_to(dest, src)
+end
+
+function MOI.supports_add_constrained_variables(::Optimizer, ::Type{<:VectorizedManifold})
+    return true
+end
+
+function MOI.add_constrained_variables(model::Optimizer, set::VectorizedManifold)
+    F = MOI.VectorOfVariables
+    if !isnothing(model.manifold)
+        throw(
+            AddConstraintNotAllowed{F,typeof(set)}(
+                "Only one manifold allowed, variables in `$(model.manifold)` have already been added.",
+            ),
+        )
+    end
+    model.manifold = set.manifold
+    n = MOI.dimension(set)
+    model.num_variables = n
+    v = MOI.VariableIndex.(1:n)
+    for _ in 1:n
+        push!(model.variable_primal_start, nothing)
+    end
+    return v, MOI.ConstraintIndex{F,typeof(set)}(1)
+end
+
+function MOI.is_valid(model::Optimizer, vi::MOI.VariableIndex)
+    return 1 <= vi.value <= model.num_variables
+end
+
+function MOI.supports(::Optimizer, ::MOI.VariablePrimalStart, ::Type{MOI.VariableIndex})
+    return true
+end
+
+function MOI.set(
+    model::Optimizer,
+    ::MOI.VariablePrimalStart,
+    vi::MOI.VariableIndex,
+    value::Union{Real,Nothing},
+)
+    MOI.throw_if_not_valid(model, vi)
+    model.variable_primal_start[vi.value] = value
+    return nothing
+end
+
+MOI.supports(::Optimizer, ::MOI.NLPBlock) = true
+
+function MOI.set(model::Optimizer, ::MOI.NLPBlock, nlp_data::MOI.NLPBlockData)
+    model.nlp_data = nlp_data
+    return nothing
+end
+
+function MOI.supports(
+    ::Optimizer, ::Union{MOI.ObjectiveSense,MOI.ObjectiveFunction{F}}
+) where {F<:_FUNCTIONS}
+    return true
+end
+
+function MOI.set(model::Optimizer, ::MOI.ObjectiveSense, sense::MOI.OptimizationSense)
+    model.sense = sense
+    return nothing
+end
+
+function MOI.get(
+    model::Optimizer, attr::Union{MOI.ObjectiveFunctionType,MOI.ObjectiveFunction}
+)
+    return MOI.get(model.qp_data, attr)
+end
+
+function MOI.set(
+    model::Optimizer, attr::MOI.ObjectiveFunction{F}, func::F
+) where {F<:_FUNCTIONS}
+    MOI.set(model.qp_data, attr, func)
+    return nothing
+end
+
+function MOI.eval_objective(model::Optimizer, x)
+    if model.sense == MOI.FEASIBILITY_SENSE
+        return 0.0
+    elseif model.nlp_data.has_objective
+        return MOI.eval_objective(model.nlp_data.evaluator, x)
+    end
+    return MOI.eval_objective(model.qp_data, x)
+end
+
+function MOI.eval_objective_gradient(model::Optimizer, grad, x)
+    if model.sense == MOI.FEASIBILITY_SENSE
```

Comment:
Similarly we have `get_gradient(problem, x)` or `get_gradient(manifold, objective, x)` but also mutating/inplace variants `get_gradient(manifold, gradient, objective, x)` for example (even independent whether the user implemented a mutating or an allocating gradient function).

Reply:
And how to you communicate the objective to the problem ?

Reply:
I do not completely understand. In Manopt (and maybe we have different things in mind when we speak of problem and objective) the objective is a struct containing all “knowledge” about the objective: the cost, maybe the gradient, maybe the Hessian, maybe some proximal maps,... all these functions have the manifold as first parameter (the cost is of the form `f(M,x)`.
the problem consists of a specific manifold `M` and the objective.
This way one can for example change the metric on the manifold without having to redefine the whole objective – or even take a completely different manifold, if the objective makes sense thereon.

Usually (that is all solvers but two) the problem hence has two fields: `M` and `objective`,
and basically `get_cost(problem, x)` just dispatches on `get_cost(problem.M, problem.objective, x)`.

 So maybe the `Manopt` `problem` is what MOI calls model?

Reply:
Ok, so we should replace
```julia
Manopt.gradient_descent(
    model.manifold, eval_f_cb, eval_grad_f_cb, start
)
```
by something like
```julia
objective = Objective(eval_f_cb, eval_grad_f_cb)
problem = Problem(model.manifold, objective)
Manopt.gradient_descent(problem, start)
```
Maybe we can even store `Problem` inside the `Optimizer` struct directly and build the callbacks whenever the objective changes.

Reply:
good ideas – nearly ;)

we have different types of objectives, depending on what they provide. for gradient descent the best would be the `ManifoldGradientObjective` ( https://manoptjl.org/stable/plans/objective/#Manopt.ManifoldGradientObjective ) and note that the gradient function can be allocating or mutating (in-place).

For now the problem got the (a little long) name `DefaultManoptProblem`, see https://manoptjl.org/stable/plans/problem/#Manopt.DefaultManoptProblem (and two solvers have a problem where they actually require 2 manifolds, that is maybe too advanced here for a first approach).

The last line – again nearly – I was for now not yet convinced enough for the line you write so there are for now
1. the “easy access” one `gradient_descent(M, f, grad_f, start)
2. the “mid layer one” `gradient_descent(M, objective, start)` (and yes one could here also do one with the problem upfront) - because 
3. the kind of _lowest_level_ veriant is: use the `GradientDescentState` ( https://manoptjl.org/stable/solvers/gradient_descent/#Manopt.GradientDescentState), where any state _uniquely_ identifies a solver and contains all its setup/options. All states have an “easy” constructor by just passing the  `manifold` to initialise them. With `problem` and `state` you can also just call `Manopt.solve!(problem, state)` which modifies state.

This is even meant in that level of granularity: 1) beginner level, easy to call 2) intermediate level, still easy to call 3) advanced, but much more direct control.
So sure we could store `problem` in there with callbacks – good idea. But when the user has decided for an algorithm we could also store the `state`.

And thanks again for all the interest and questions. I had MOI in sight for a while but only read parts of the docs until now and never felt Manopt would actually fit really in there. So great to explore whether it does :)

Reply:
You're welcome ;) Yes, the lowest_level seems to be the way to go here. No problem with long names, we took that road as well with JuMP/MOI